### ISSUE TITLE
[Bug]: Use vllm to obtain Qwen3VL last token hidden status

### ISSUE THREAD

**Issue by JasonLeeUT** [vllm-project/vllm#32858]
### Your current environment

from hdfs_io import hopen, hexists, hlist_files,hmv
from qwen_vl_utils.vision_process import smart_resize

import torch, base64, io
from dataclasses import asdict
import json
import os
import argparse
import torch
import base64
import re
from PIL import Image
from io import BytesIO
import numpy as np
# from vllm import LLM, SamplingParams
from vllm import LLM, EngineArgs
import requests
from base64 import b64decode
import sys
import random
from tqdm import tqdm
from concurrent.futures import ThreadPoolExecutor, as_completed
import os
os.environ['VLLM_WORKER_MULTIPROC_METHOD'] = 'spawn'
# 或者使用 spawn 模式
import torch.multiprocessing as mp
mp.set_start_method('spawn', force=True)

PIXEL_SIZE = 100 * 32 * 32
SIZE_FACTOR = 32
TXT_MAX_LEN = 1024

base_template = "<|im_start|>system\n用一个词总结以下视频和文字.<|im_end|>\n<|im_start|>user\n<|vision_start|><|image_pad|><|vision_end|>{}\n<|im_end|>\n<|im_start|>assistant\n<|endoftext|>"

# base_template_text = "用一个词总结这句话:\n{}\n<|file_sep|>"
base_template_text = "<|im_start|>system\n用一个词总结以下文字.<|im_end|>\n<|im_start|>user\n{}<|im_end|>\n<|im_start|>assistant\n<|endoftext|>"

def qwen2_resize_image(image, min_pixels=PIXEL_SIZE, max_pixels=PIXEL_SIZE):
    """调整图像尺寸以适应模型输入要求"""
    # image = to_rgb(image)
    width, height = image.size
    resized_height, resized_width = smart_resize(
        height,
        width,
        factor=SIZE_FACTOR,
        min_pixels=min_pixels,
        max_pixels=max_pixels,
    )
    return image.resize((resized_width, resized_height))

# 辅助函数：处理单张图像（缩放，供线程池调用）
def process_image(img_bytes, j):
    img = Image.open(BytesIO(img_bytes))
    resized_img = qwen2_resize_image(img)
    return resized_img

def toImage(image_binary):
    image_stream = io.BytesIO(image_binary)
    return Image.open(image_stream)

def get_image(images):
    resized_image = [qwen2_resize_image(toImage(image)) for image in images]
    return resized_image
    
def get_prompt(title, num_image):
    return base_template.format(title).replace('<|vision_start|><|image_pad|><|vision_end|>', '<|vision_start|><|image_pad|><|vision_end|>' * num_image)

def get_prompt_text(title):
    return base_template_text.format(title)

def get_query(title, image_urls):
    # if type(title) == bytes:
    #     title = title.decode()
    if image_urls is not None and len(image_urls) > 0:
        images = get_images_from_urls(image_urls)
        # image_binary = ast.literal_eval(image_binary.decode())
        # data = get_image(image_binary)
        prompt = get_prompt(title[: TXT_MAX_LEN], len(images))
        # print (prompt)
        return {'prompt': prompt, 'multi_modal_data': {'image': images}}
    else:
        prompt = get_prompt_text(title[: TXT_MAX_LEN ])
        return {'prompt': prompt}
    
def encode_str(s, flag=True):
    if flag:
        return s.encode()
    else:
        return s

def main():
    # 解析命令行参数
    parser = argparse.ArgumentParser()
    
    parser.add_argument("--num-machines", type=int, default=1)
    parser.add_argument("--machine-id", type=int, default=0)
    parser.add_argument("--num-gpus", type=int, default=1)
    parser.add_argument('--process-idx', type=int, default=0)
    parser.add_argument("--batch-size", type=int, default=32) #64

    args = parser.parse_args()

    # 提取参数
        
    num_machines = args.num_machines
    machine_id = args.machine_id
    num_gpus = args.num_gpus
    process_idx = args.process_idx
    batch_size = args.batch_size

    # 初始化LLM模型
    # checkpoint_path = "./2B-1yi-sota-for-vllm"
    checkpoint_path="./Qwen3-VL-2B-Instruct"

    engine_args = EngineArgs(
        model=checkpoint_path,
        dtype='float32',  # 改为float32，减少bfloat16的精度波动
        # dtype='bfloat16',
        # task="embed",
        runner='pooling',
        trust_remote_code=True,
        limit_mm_per_prompt={"image": 8, "video": 0},
        gpu_memory_utilization=0.8, #0.8
        max_model_len=2560, #这里的是分词后的token长度，所以我们可能超，因为1024个字符，不一定等于那么多的token数
        max_num_batched_tokens=2560*batch_size,
        seed=42,
        enforce_eager=False
    )
    e = asdict(engine_args)
    llm = LLM(**e)
    
    print("llm initialized")

    if num_machines==1:
        machine_id=0 # 避免出问题，单机程序
    if machine_id >= num_machines:
        machine_id = random.randint(0, num_machines - 1)  # 随机取 0 到 num_machines-1 的整数
        print(f"machine_id 超出有效范围（0~{num_machines-1}），已随机分配为: {machine_id}")
    print(f"num_machines: {num_machines}")
    print(f"num_gpus: {num_gpus}")
    print(f"process_idx: {process_idx}")
    print(f"batch_size: {batch_size}")
    print(f"machine_id: {machine_id}")
    # 先计算总数
    total_files = 0
    idx = 0
    
    hdfs_files_gpu = ['/mnt/bn/ljf-lq/qwen3_code/1225bestconfig/Qwen3-vl-embedding-caption-bad-cases-collect/cases/上海旅游攻略.json']

    # 处理文件
    idx = 0

    # 创建进度条
    pbar = tqdm(total=total_files)
    # hdfs_files_gpu = ["hdfs://haruna/home/byte_search_recall/user/jianfengli/L1-back5yi-b64save/00000/part-19676.zstd"] # debug
    for file in hdfs_files_gpu:

        file_name = os.path.basename(file)
        part_name = os.path.splitext(file_name)[0]

        prefix = file.split('/')[-2]

        print(f"处理文件: {file}")
        # 处理当前文件
        try:
            with hopen(file, 'r') as f:
                    dataj = json.load(f)
                    inputs = []
                    for data in dataj:
                      

                        title=str(data.get('title','')).replace(' ', '').replace('\t', '').replace('\n', '').strip()
                        ocr=str(data.get('ocr','')).replace('\t', '').replace('\n', '').strip()
                        poi=str(data.get('poi','')).replace(' ', '').replace('\t', '').replace('\n', '').strip()
                        username=str(data.get('username','')).replace(' ', '').replace('\t', '').replace('\n', '').strip()
                        music=str(data.get('music','')).replace(' ', '').replace('\t', '').replace('\n', '').strip()
                        text = f"\n标题：{title.strip()}\n字幕：{ocr.strip()}\n位置：{poi.strip()}\n用户名：{username.strip()}\n音乐：{music.strip()}"
                        # text = f"\n标题：\n字幕：\n位置：\n用户名：\n音乐："
                        if len(data.get('image_b64', []))>0:
                            images=[Image.open(BytesIO(base64.b64decode(img))) for img in data['image_b64']] 
                            prompt = get_prompt(text[: TXT_MAX_LEN], len(images))
                            cur_input= {'prompt': prompt, 'multi_modal_data': {'image': images}}
                            inputs.append(cur_input)
                        else:
                            prompt = get_prompt_text(text[: TXT_MAX_LEN ])
                            cur_input= {'prompt': prompt}
                            inputs.append(cur_input)

                    outputs = llm.embed(inputs)

                    embeddings = []
                    for out in outputs:
                        # 转为tensor并确保是float类型（避免整数类型影响点积）
                        emb_tensor = torch.tensor(out.outputs.embedding, dtype=torch.float32)
                        # 确保是1维（若为多维需展平，根据你的Embedding格式调整）
                        emb_tensor = emb_tensor.flatten()
                        embeddings.append(emb_tensor)

                    # 拼接成2维矩阵：shape [n, dim]
                    emb_matrix = torch.stack(embeddings, dim=0)

                    # 步骤2：计算两两之间的点积矩阵（shape: [n, n]）
                    # 矩阵乘法：emb_matrix @ emb_matrix.T → 第i行第j列 = emb_i 和 emb_j 的点积
                    dot_product_matrix = torch.matmul(emb_matrix, emb_matrix.T)

                    # ===================== 输出结果 =====================
                    print("点积矩阵 shape：", dot_product_matrix.shape)  # 输出：torch.Size([3, 3])（若有3个Embedding）
                    print("点积矩阵：")
                    print(dot_product_matrix.numpy())  # 转为numpy更易阅读

                    # # （可选）若需要归一化后的点积（余弦相似度），补充以下代码
                    # emb_matrix_norm = torch.nn.functional.normalize(emb_matrix, p=2, dim=-1)
                    # cos_sim_matrix = torch.matmul(emb_matrix_norm, emb_matrix_norm.T)
                    # print("\n归一化后的点积矩阵（余弦相似度）：")
                    # print(cos_sim_matrix.numpy())

                    # emb0 = torch.nn.functional.normalize(torch.tensor(outputs[0].outputs.embedding), p=2, dim=-1)
                    # emb1 = torch.nn.functional.normalize(torch.tensor(outputs[1].outputs.embedding), p=2, dim=-1)
                    # emb2 = torch.nn.functional.normalize(torch.tensor(outputs[2].outputs.embedding), p=2, dim=-1)
                    # dot01 = torch.dot(emb0, emb1)
                    # dot02 = torch.dot(emb0, emb2)
                    # dot12 = torch.dot(emb1, emb2)
                    # print(f"dot01: {dot01}, dot02: {dot02}, dot12: {dot12}")
                    # data['embedding0'] = emb0.tolist()
                    # data['embedding1'] = emb1.tolist()
                    # data['embedding2'] = emb2.tolist()
                    # data['dot01'] = dot01.item()
                    # data['dot02'] = dot02.item()
                    # data['dot12'] = dot12.item()
            
            new_json_path = file[:-5]+'-testfloat32-3-eagerclose-noembed.json'
            try:
                # 2. 写入新JSON文件（核心操作）
                with open(new_json_path, "w", encoding="utf-8") as f:
                    # ensure_ascii=False：保留中文；indent=4：格式化输出（可读性强）
                    # default=str：兼容少量非JSON原生类型（如numpy数值，可选）
                    json.dump(
                        dataj,
                        f,
                        ensure_ascii=False,
                        indent=4,
                        default=str  # 可选：处理极少数非标准类型（如torch.Tensor需先转list）
                    )
                print(f"修改后的dataj已保存为：{new_json_path}")
            except Exception as e:
                print(f"保存JSON失败：{e}")

        except Exception as e:
            print(e)
            inputs = []
            gid_list = []
if __name__ == '__main__':
    # 解决多进程启动问题的关键保护
    main() 

I have above file to get embedding using vllm. I have finetuned a model from Qwen3VL and I use the hidden status of the last token as my embedding. However, the embedding I got from vllm is greatly different from transformers. The dot of two embeddings can be 0.6 after normalization. Does anyone have similiar issues related to qwen3VL?

### 🐛 Describe the bug

Vllm Qwen3VL Embedding bug.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.
